Fix crew upgrade, llvm => 18.1.3, upx => 4.2.3, update ca-certificates

satmandu opened the pull request:
Fixes #9642  by adding a missing `?` to `Package.load_package(pkg_file, pkg_name).is_fake`.

- Also fix verbose `crew remove`:
```bash
crew reinstall -v upx
...
Removing file /usr/local/share/man/man1/upx.1.zst
Removing directory /usr/local/share/doc/upx
/usr/local/lib/crew/commands/remove.rb:57:in `remove': undefined local variable or method `pkg_name' for class Command (NameError)

    puts "Removing package #{pkg_name} from device.json".yellow if verbose
                             ^^^^^^^^
        from /usr/local/bin/crew:1258:in `install'
        from /usr/local/bin/crew:1109:in `resolve_dependencies_and_install'
        from /usr/local/bin/crew:1791:in `block in reinstall_command'
        from /usr/local/bin/crew:1781:in `each'
        from /usr/local/bin/crew:1781:in `reinstall_command'
        from /usr/local/bin/crew:1878:in `<main>'
```

Works properly:
- [x] `x86_64`
- [x] `i686`
- [x] `armv7l` <!-- (reasons why it doesn't) -->

### Run the following to get this pull request's changes locally for testing.
```
CREW_REPO=https://github.com/satmandu/chromebrew.git CREW_BRANCH=llvm crew update ; yes | crew upgrade
```

<!--
## That's it
Thank you for submitting your pull request.
When done, please delete the parts of this template which you don't need or these, which are only for guidance.
-->

Zopolis4 commented on lib/package.rb:
```diff
@@ -32,7 +32,7 @@ def self.print_boolean_properties
 
   class << self
-    attr_accessor :name, :cached_build, :in_build, :build_from_source, :in_upgrade
+    attr_accessor :name, :cached_build, :in_build, :build_from_source, :in_upgrade, :arch_flags_override, :conflicts_ok, :git_clone_deep, :git_fetchtags, :gnome, :is_fake, :is_musl, :is_static, :no_compile_needed, :no_compress, :no_env_options, :no_fhs, :no_git_submodules, :no_links, :no_lto, :no_patchelf, :no_shrink, :no_source_build, :no_strip, :no_upstream_update, :no_zstd, :patchelf, :print_s ...
```
Ah, I see. In that case, could we only add the relevant properties?

satmandu replied:
The boolean properties *such as `@pkg.is_fake` don't seem to be available after `Command.remove(@pkg, @opt_verbose)` is run without this.

Is there a better solution?

Zopolis4 replied:
Command.remove doesn't even use is_fake... 

@supechicken, any idea what's going on?

satmandu replied:
Agree on only including the properties we need there, but I haven't figured out why things are broken, and hence adding them all seemed to be the conservative choice.

satmandu replied:
Also, it appears that @pkg.is_fake is already broken before we get to `Command.remove(@pkg, @opt_verbose)` as putting this before that line (and using the original `package.rb`) shows that boolean property to be true there too:
```ruby
 puts "The #{@pkg.name} is_fake property is #{@pkg.is_fake}.".orange
```

![image](https://github.com/chromebrew/chromebrew/assets/1096701/db809dbe-040c-4df1-9105-cf4153e067aa)

uberhacker replied:
Shouldn't the tests be catching these issues prior to submitting a PR? :stuck_out_tongue_winking_eye:

satmandu replied:
We need a test for each of these...

satmandu replied:
Also, it looks like upgrade may not be seeing `@pkg.is_fake` at all...

satmandu replied:
Drop this in `crew` and then just put `crewlog` in places to have `@pkg.is_fake` logged to stdout:
```ruby
def crewlog
  str = "The #{@pkg.name} is_fake property is #{@pkg.is_fake}."
  caller_locations(1, 1).first.tap {|loc| puts "#{loc.path}:#{loc.lineno}:#{caller_locations(3).first.label}:#{str}".lightpurple if @opt_verbose}
end
```

satmandu replied:
Turns out this doesn't actually fix it... it just sets it to false.

We're not reading this from the file properly...

satmandu replied:
All these changes have been removed...

uberhacker commented on bin/crew:
```diff
@@ -1258,7 +1258,9 @@ def install
 
-  unless @pkg.is_fake?
+  if @pkg.is_fake?
+    puts "The #{@pkg.name} is_fake property is true. File installs will be skipped.".orange if @opt_verbose
```
This wording is technically correct but I'm thinking it might be more human readable to say:
```
put "Install will be skipped since #{@pkg.name} is a fake package."...
```

uberhacker commented on packages/ca_certificates.rb:
```diff
@@ -123,6 +123,10 @@ def self.install
+    if File.file?("#{CREW_PREFIX}/bin/update-ca-certificates")
+      system "#{CREW_PREFIX}/bin/update-ca-certificates --fresh --certsconf #{CREW_PREFIX}/etc/ca-certificates.conf"
+    else
+      puts "#{CREW_PREFIX}/bin/update-ca-certificates is missing!".lightred
```
Provide a suggestion how to fix too?

uberhacker replied:
Wouldn't a dependency on the package that includes the update-ca-certificates binary be the solution to fix this?

uberhacker replied:
Wait...the executable is from this package? :confused:

satmandu replied:
Yeah that's how I realized crew upgrade was broken...

uberhacker commented on bin/crew:
```diff
@@ -115,6 +115,11 @@ at_exit do
   ExitMessage.print
 end
 
+def crewlog
```
Would it be better to pass the message as a parameter?  I know this will require the message to be created outside the function but it would make it more flexible.  Right now, this generic named function is only logging for is_fake.

satmandu replied:
I only have this in there for messing around to figure out what is going on. I will remove it later.

satmandu commented:
For future use, this is the code block I was using for debugging:
```ruby
def crewlog
  str = "#{@pkg.name} is_fake property:  #{@pkg.is_fake?}, no_source_build:  #{@pkg.no_source_build?}."
  caller_locations(1, 1).first.tap {|loc| puts "#{loc.path}:#{loc.lineno}:#{caller_locations(3).first.label}:#{str}".lightpurple if @opt_verbose}
end
```

Zopolis4 commented:
Sorry about missing the question mark on the is_fake when I updated the code-- Once I split out the upgrade function (and its dependencies on build and install and their dependencies) we can get proper testing for that.

satmandu commented:
> Sorry about missing the question mark on the is_fake when I updated the code-- Once I split out the upgrade function (and its dependencies on build and install and their dependencies) we can get proper testing for that.

I'm just glad we figured it out!

We definitely need unit tests for everything... Maybe we don't replace a function by splitting it out unless we add a unit test for that function? 😅

Zopolis4 commented:
I can't add unit tests for some functions until I split out others (i.e. I can't test removing packages because we can't install them through calling a function)
